File: core/src/test/scala/unit/kafka/server/CreateTopicsRequestWithPolicyTest.scala
/**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
  * this work for additional information regarding copyright ownership.
  * The ASF licenses this file to You under the Apache License, Version 2.0
  * (the "License"); you may not use this file except in compliance with
  * the License.  You may obtain a copy of the License at
  *
  *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

package kafka.server

import java.util
import java.util.Properties
import org.apache.kafka.common.config.TopicConfig
import org.apache.kafka.common.errors.PolicyViolationException
import org.apache.kafka.common.internals.Topic
import org.apache.kafka.common.protocol.Errors
import org.apache.kafka.server.config.ServerLogConfigs.CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG
import org.apache.kafka.server.policy.CreateTopicPolicy
import org.apache.kafka.server.policy.CreateTopicPolicy.RequestMetadata
import org.junit.jupiter.api.TestInfo
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

import scala.jdk.CollectionConverters._

class CreateTopicsRequestWithPolicyTest extends AbstractCreateTopicsRequestTest {
  import CreateTopicsRequestWithPolicyTest._

  override def brokerPropertyOverrides(properties: Properties): Unit = {
    super.brokerPropertyOverrides(properties)
    properties.put(CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG, classOf[Policy].getName)
  }

  override def kraftControllerConfigs(testInfo: TestInfo): Seq[Properties] = {
    val properties = new Properties()
    properties.put(CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG, classOf[Policy].getName)
    Seq(properties)
  }

  @ParameterizedTest
  @ValueSource(strings = Array("kraft"))
  def testValidCreateTopicsRequests(quorum: String): Unit = {
    validateValidCreateTopicsRequests(topicsReq(Seq(topicReq("topic1",
      numPartitions = 5))))

    validateValidCreateTopicsRequests(topicsReq(Seq(topicReq("topic2",
      numPartitions = 5, replicationFactor = 3)),
      validateOnly = true))

    validateValidCreateTopicsRequests(topicsReq(Seq(topicReq("topic3",
      numPartitions = 11, replicationFactor = 2,
      config = Map(TopicConfig.RETENTION_MS_CONFIG -> 4999.toString))),
      validateOnly = true))

    validateValidCreateTopicsRequests(topicsReq(Seq(topicReq("topic4",
      assignment = Map(0 -> List(1, 0), 1 -> List(0, 1))))))
  }

  @ParameterizedTest
  @ValueSource(strings = Array("kraft"))
  def testErrorCreateTopicsRequests(quorum: String): Unit = {
    val existingTopic = "existing-topic"
    createTopic(existingTopic, 5)

    // Policy violations
    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("policy-topic1",
      numPartitions = 4, replicationFactor = 1))),
      Map("policy-topic1" -> error(Errors.POLICY_VIOLATION, Some("Topics should have at least 5 partitions, received 4"))))

    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("policy-topic2",
      numPartitions = 4, replicationFactor = 3)), validateOnly = true),
      Map("policy-topic2" -> error(Errors.POLICY_VIOLATION, Some("Topics should have at least 5 partitions, received 4"))))

    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("policy-topic3",
      numPartitions = 11, replicationFactor = 2,
      config = Map(TopicConfig.RETENTION_MS_CONFIG -> 5001.toString))), validateOnly = true),
      Map("policy-topic3" -> error(Errors.POLICY_VIOLATION,
        Some("RetentionMs should be less than 5000ms if replicationFactor > 5"))))

    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("policy-topic4",
      numPartitions = 11, replicationFactor = 3,
      config = Map(TopicConfig.RETENTION_MS_CONFIG -> 5001.toString))), validateOnly = true),
      Map("policy-topic4" -> error(Errors.POLICY_VIOLATION,
        Some("RetentionMs should be less than 5000ms if replicationFactor > 5"))))

    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("policy-topic5",
      assignment = Map(0 -> List(1), 1 -> List(0)),
      config = Map(TopicConfig.RETENTION_MS_CONFIG -> 5001.toString))), validateOnly = true),
      Map("policy-topic5" -> error(Errors.POLICY_VIOLATION,
        Some("Topic partitions should have at least 2 partitions, received 1 for partition 0"))))

    // Check that basic errors still work
    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq(existingTopic,
      numPartitions = 5, replicationFactor = 1))),
      Map(existingTopic -> error(Errors.TOPIC_ALREADY_EXISTS,
        Some("Topic 'existing-topic' already exists."))))

    var errorMsg = "Unable to replicate the partition 4 time(s): The target replication factor of 4 cannot be reached because only 3 broker(s) are registered."
    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("error-replication",
      numPartitions = 10, replicationFactor = brokerCount + 1)), validateOnly = true),
      Map("error-replication" -> error(Errors.INVALID_REPLICATION_FACTOR,
        Some(errorMsg))))

    errorMsg = "Replication factor must be larger than 0, or -1 to use the default value."
    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("error-replication2",
      numPartitions = 10, replicationFactor = -2)), validateOnly = true),
      Map("error-replication2" -> error(Errors.INVALID_REPLICATION_FACTOR,
        Some(errorMsg))))

    errorMsg = "Number of partitions was set to an invalid non-positive value."
    validateErrorCreateTopicsRequests(topicsReq(Seq(topicReq("error-partitions",
      numPartitions = -2, replicationFactor = 1)), validateOnly = true),
      Map("error-partitions" -> error(Errors.INVALID_PARTITIONS,
        Some(errorMsg))))
  }

}

object CreateTopicsRequestWithPolicyTest {

  class Policy extends CreateTopicPolicy {

    var configs: Map[String, _] = _
    var closed = false

    def configure(configs: util.Map[String, _]): Unit = {
      this.configs = configs.asScala.toMap
    }

    def validate(requestMetadata: RequestMetadata): Unit = {
      if (Topic.isInternal(requestMetadata.topic())) {
        // Do not verify internal topics
        return
      }
      require(!closed, "Policy should not be closed")
      require(configs.nonEmpty, "configure should have been called with non empty configs")

      import requestMetadata._
      if (numPartitions != null || replicationFactor != null) {
        require(numPartitions != null, s"numPartitions should not be null, but it is $numPartitions")
        require(replicationFactor != null, s"replicationFactor should not be null, but it is $replicationFactor")
        require(replicasAssignments == null, s"replicaAssignments should be null, but it is $replicasAssignments")

        if (numPartitions < 5)
          throw new PolicyViolationException(s"Topics should have at least 5 partitions, received $numPartitions")

        if (numPartitions > 10) {
          if (requestMetadata.configs.asScala.get(TopicConfig.RETENTION_MS_CONFIG).fold(true)(_.toInt > 5000))
            throw new PolicyViolationException("RetentionMs should be less than 5000ms if replicationFactor > 5")
        } else
          require(requestMetadata.configs.isEmpty, s"Topic configs should be empty, but it is ${requestMetadata.configs}")

      } else {
        require(numPartitions == null, s"numPartitions should be null, but it is $numPartitions")
        require(replicationFactor == null, s"replicationFactor should be null, but it is $replicationFactor")
        require(replicasAssignments != null, s"replicaAssignments should not be null, but it is $replicasAssignments")

        replicasAssignments.asScala.toSeq.sortBy { case (tp, _) => tp }.foreach { case (partitionId, assignment) =>
          if (assignment.size < 2)
            throw new PolicyViolationException("Topic partitions should have at least 2 partitions, received " +
              s"${assignment.size} for partition $partitionId")
        }
      }

    }

    def close(): Unit = closed = true

  }
}

File: clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/CreateTopicsRequestWithPolicyTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.kafka.clients;

import org.apache.kafka.clients.admin.Admin;
import org.apache.kafka.clients.admin.CreateTopicsOptions;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.common.config.TopicConfig;
import org.apache.kafka.common.errors.InvalidPartitionsException;
import org.apache.kafka.common.errors.InvalidReplicationFactorException;
import org.apache.kafka.common.errors.PolicyViolationException;
import org.apache.kafka.common.errors.TopicExistsException;
import org.apache.kafka.common.internals.Topic;
import org.apache.kafka.common.test.ClusterInstance;
import org.apache.kafka.common.test.api.ClusterConfigProperty;
import org.apache.kafka.common.test.api.ClusterTest;
import org.apache.kafka.common.test.api.ClusterTestDefaults;
import org.apache.kafka.server.config.ServerLogConfigs;
import org.apache.kafka.server.policy.CreateTopicPolicy;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@ClusterTestDefaults(
    brokers = 3,
    serverProperties = {
        @ClusterConfigProperty(key = ServerLogConfigs.CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG, value = "org.apache.kafka.clients.CreateTopicsRequestWithPolicyTest$Policy"),
    }
)
public class CreateTopicsRequestWithPolicyTest {

    public static class Policy implements CreateTopicPolicy {
        private Map<String, ?> configs;
        private boolean closed = false;

        @Override
        public void configure(Map<String, ?> configs) {
            this.configs = configs;
        }

        @Override
        public void validate(RequestMetadata requestMetadata) throws PolicyViolationException {
            if (Topic.isInternal(requestMetadata.topic())) {
                // Do not verify internal topics
                return;
            }

            if (closed) {
                throw new IllegalStateException("Policy should not be closed");
            }
            if (configs == null || configs.isEmpty()) {
                throw new IllegalStateException("Configure should have been called with non empty configs");
            }

            if (requestMetadata.numPartitions() != null || requestMetadata.replicationFactor() != null) {
                assertNotNull(requestMetadata.numPartitions(), "numPartitions should not be null, but it is " + requestMetadata.numPartitions());
                assertNotNull(requestMetadata.replicationFactor(), "replicationFactor should not be null, but it is " + requestMetadata.replicationFactor());
                assertNull(requestMetadata.replicasAssignments(), "replicaAssignments should be null, but it is " + requestMetadata.replicasAssignments());

                if (requestMetadata.numPartitions() < 5) {
                    throw new PolicyViolationException("Topics should have at least 5 partitions, received " +
                            requestMetadata.numPartitions());
                }

                if (requestMetadata.numPartitions() > 10) {
                    String retentionMs = requestMetadata.configs().get(TopicConfig.RETENTION_MS_CONFIG);
                    if (retentionMs == null || Integer.parseInt(retentionMs) > 5000) {
                        throw new PolicyViolationException("RetentionMs should be less than 5000ms if partitions > 10");
                    }
                } else {
                    assertTrue(requestMetadata.configs().isEmpty(), "Topic configs should be empty, but it is " + requestMetadata.configs());
                }
            } else {
                assertNull(requestMetadata.numPartitions(), "numPartitions should be null, but it is " + requestMetadata.numPartitions());
                assertNull(requestMetadata.replicationFactor(), "replicationFactor should be null, but it is " + requestMetadata.replicationFactor());
                assertNotNull(requestMetadata.replicasAssignments(), "replicasAssignments should not be null, but it is " + requestMetadata.replicasAssignments());

                requestMetadata.replicasAssignments().forEach((partitionId, replicas) -> {
                    if (replicas.size() < 2) {
                        throw new PolicyViolationException("Topic partitions should have at least 2 replicas, received " +
                                replicas.size() + " for partition " + partitionId);
                    }
                });
            }
        }

        @Override
        public void close() {
            closed = true;
        }
    }

    private void validateValidCreateTopicsRequests(NewTopic topic, Admin admin, boolean validateOnly) throws Exception {
        admin.createTopics(
            List.of(topic),
            new CreateTopicsOptions().validateOnly(validateOnly)
        ).all().get();
    }

    @ClusterTest
    public void testValidCreateTopicsRequests(ClusterInstance cluster) throws Exception {
        try (Admin admin = cluster.admin()) {
            cluster.createTopic("topic1", 5, (short) 1);

            validateValidCreateTopicsRequests(
                new NewTopic("topic2", 5, (short) 3),
                admin,
                true
            );

            validateValidCreateTopicsRequests(
                new NewTopic("topic3", 11, (short) 2)
                    .configs(Map.of(TopicConfig.RETENTION_MS_CONFIG, "4999")),
                admin,
                true
            );

            validateValidCreateTopicsRequests(
                new NewTopic("topic4", Map.of(
                    0, List.of(1, 0),
                    1, List.of(0, 1)
                )),
                admin,
                false
            );
        }
    }

    private void validateErrorCreateTopicsRequests(NewTopic topic, Admin admin, boolean validateOnly, Class<? extends Throwable> expectedExceptionClass, String expectedErrorMessage) {
        ExecutionException exception = assertThrows(ExecutionException.class, () ->
                admin.createTopics(List.of(topic), new CreateTopicsOptions().validateOnly(validateOnly)).all().get());
        assertEquals(
                expectedExceptionClass,
                exception.getCause().getClass(),
                "Expected " + expectedExceptionClass.getSimpleName() + ", but got " + exception.getCause().getClass().getSimpleName()
        );
        assertTrue(exception.getMessage().contains(expectedErrorMessage));
    }

    @ClusterTest
    public void testErrorCreateTopicsRequests(ClusterInstance cluster) throws Exception {
        try (Admin admin = cluster.admin()) {
            String existingTopic = "existing-topic";
            cluster.createTopic(existingTopic, 5, (short) 1);

            // Policy violations
            validateErrorCreateTopicsRequests(
                new NewTopic("policy-topic1", 4, (short) 1),
                admin,
                false,
                PolicyViolationException.class,
                "Topics should have at least 5 partitions, received 4"
            );

            validateErrorCreateTopicsRequests(
                new NewTopic("policy-topic2", 4, (short) 3),
                admin,
                true,
                PolicyViolationException.class,
                "Topics should have at least 5 partitions, received 4"
            );

            validateErrorCreateTopicsRequests(
                new NewTopic("policy-topic3", 11, (short) 2)
                    .configs(Map.of(TopicConfig.RETENTION_MS_CONFIG, "5001")),
                admin,
                true,
                PolicyViolationException.class,
                "RetentionMs should be less than 5000ms if partitions > 10"
            );

            validateErrorCreateTopicsRequests(
                new NewTopic("policy-topic4", Map.of(
                    0, List.of(1),
                    1, List.of(0)
                )).configs(Map.of(TopicConfig.RETENTION_MS_CONFIG, "5001")),
                admin,
                true,
                PolicyViolationException.class,
                "Topic partitions should have at least 2 replicas, received 1 for partition 0"
            );

            // Check that basic errors still work
            validateErrorCreateTopicsRequests(
                new NewTopic(existingTopic, 5, (short) 1),
                admin,
                false,
                TopicExistsException.class,
                "Topic 'existing-topic' already exists."
            );

            validateErrorCreateTopicsRequests(
                new NewTopic("error-replication", 10, (short) 4),
                admin,
                true,
                InvalidReplicationFactorException.class,
                "Unable to replicate the partition 4 time(s): The target replication factor of 4 cannot be "
                    + "reached because only 3 broker(s) are registered or some brokers have all their log directories cordoned."
            );

            validateErrorCreateTopicsRequests(
                new NewTopic("error-replication2", 10, (short) -2),
                admin,
                true,
                InvalidReplicationFactorException.class,
                "Replication factor must be larger than 0, or -1 to use the default value."
            );

            validateErrorCreateTopicsRequests(
                new NewTopic("error-partitions", -2, (short) 1),
                admin,
                true,
                InvalidPartitionsException.class,
                "Number of partitions was set to an invalid non-positive value."
            );
        }
    }
}
